### PR TITLE
Task list component guidance and updated pattern page

### PR DESCRIPTION
**Pull request by frankieroberto**
Adds a new guidance page for the [new Task List component](https://github.com/alphagov/govuk-frontend/pull/2261).

It also updates the existing Task list page pattern, replacing the screenshot image with the new Task List component, and updating the guidance.

➡️  [Preview](https://deploy-preview-1994--govuk-design-system-preview.netlify.app/components/task-list/)

### CODE REVIEW

**Comment by paulrobertlloyd** [comment]
With regards to statuses and their relationship to the tags component, wondering if it might be desirable to have the implementation be a little more flexible, and also consistent with how other components allow for content to be provided as either `text` or `html` (for which the HTML content could be a tag):

```njk
{{ govukTaskList({
  items: [{
    title: {
      text: "Company details"
    },
    href: "#",
    status: {
      text: "Completed"
    }
  }, {
    title: {
      text: "Project details"
    },
    href: "#",
    status: {
      html: govukTag({
        text: "Incomplete",
        classes: "govuk-tag--blue",
      }),
    }
  }, {
    title: {
      text: "Fund applied to"
    },
    hint: {
      text: "The funds will be announced on 1 April 2022"
    },
    status: {
      text: "Cannot start yet",
      classes: "govuk-!-secondary-text-colour"
    }
  }]
}) }}
```

(Note: I have provided the secondary text colour in this example by using a new utility class, `govuk-!-secondary-text-colour`. If that wasn’t deemed a useful enough utility class for broader inclusion in the Design System, you could use the proposed class `govuk-task-list__status--secondary-text-colour`.)

***

Allowing for any HTML be added may be *too* flexible, and contradict guidance which is trying to direct authors to use statuses in a restricted and predictable fashion.

Also, while the above example of using a Nunjucks component as a value for a `html` parameter is possible, it’s not something ever used in any of the existing Design System guidance.

An alternative therefore might be to allow for either `status.text` or `status.tag` parameters (with `status.tag` taking precedence if both provided). This somewhat follows the design of the Phase Banner component, which also consumes the tag component and has a `tag` parameter. So for example:

```njk
{{ govukTaskList({
  items: [{
    title: {
      text: "Company details"
    },
    href: "#",
    status: {
      text: "Completed"
    }
  }, {
    title: {
      text: "Project details"
    },
    href: "#",
    status: {
      tag: {
        text: "Incomplete",
        classes: "govuk-tag--blue",
      },
    }
  }, {
    title: {
      text: "Fund applied to"
    },
    hint: {
      text: "The funds will be announced on 1 April 2022"
    },
    status: {
      text: "Cannot start yet",
      classes: "govuk-!-secondary-text-colour"
    }
  }]
}) }}
```

The HTML for the above would be something like this:

```html
<ul class="govuk-task-list">
  <li class="govuk-task-list__item govuk-task-list__item--with-link">
    <span class="govuk-task-list__task-name-and-hint">
      <a class="govuk-link govuk-task-list__link" href="#" aria-describedby="task-list-1-status">Company details</a>
    </span>
    <span class="govuk-task-list__status" id="task-list-1-status">
      Completed
    </span>
  </li>
  <li class="govuk-task-list__item govuk-task-list__item--with-link">
    <span class="govuk-task-list__task-name-and-hint">
      <a class="govuk-link govuk-task-list__link" href="#" aria-describedby="task-list-2-status">Project details</a>
    </span>
    <span class="govuk-task-list__status">
      <strong class="govuk-tag govuk-tag--blue" id="task-list-2-status">
        Incomplete
      </strong>
    </span>
  </li>
  <li class="govuk-task-list__item">
    <span class="govuk-task-list__task-name-and-hint">
      <div class="govuk-task-list__task-no-link">Fund applied to</div>
      <div id="task-list-3-hint" class="govuk-task-list__task_hint">
        The funds will be announced on 1 April 2022
      </div>
    </span>
    <span class="govuk-task-list__status govuk-!-secondary-text-colour" id="task-list-3-status">
      Cannot start yet
    </span>
  </li>
</ul>
```

**Comment by paulrobertlloyd** [comment]
Sidenote: wondering about the usefulness of the secondary text colour as an option – does it add anything? Would most users notice?

**Comment by frankieroberto** [comment]
@paulrobertlloyd all good questions! The original intention was that the task list component wouldn’t use the tag component (as the design was so different) and so we just had a “status” element within each row instead.

The working group weren’t happy with the inconsistency this would introduce though, so we’ve started exploring how to update the tag component. We’ve not yet explored how with would affect the task list implementation, but I was presuming that it would directly use the updated tag component, in the same way that the Phase banner does - ie with a `tag` key instead of `status`. Open to ideas though!

**Comment by frankieroberto** [comment]
> Sidenote: wondering about the usefulness of the secondary text colour as an option – does it add anything? Would most users notice the difference?

It’s intended only for use as “Cannot start yet”, where the whole row wouldn’t be clickable. See draft guidance: https://deploy-preview-1994--govuk-design-system-preview.netlify.app/components/task-list/

The grey is very close to the black though, admittedly!

**Comment by paulrobertlloyd** [comment]
> It’s intended only for use as “Cannot start yet”, where the whole row wouldn’t be clickable.

Oh, I get the intention. And the designer in me really likes it! Love a bit of grey text. I just wonder if in reality it really adds anything that the words don’t already convey. And yes, long history of grey text getting ever darker to meet accessibility standards, further suggesting this may be unwarranted, and a needless complexity in the grand scheme of things.

**Comment by paulrobertlloyd** [comment]
1. The Nunnucks options for the [Phase banner component](https://design-system.service.gov.uk/components/phase-banner/) show for `tag` options:

    > Options for the tag component. See [tag](https://design-system.service.gov.uk/components/tag/#options-tag-example).

    Do we want to do the same here?

2. There are options shown for `url.text`, `url.html` and `url.classes`, but going by the examples, I don’t think the `url` option exists on items. Should this not be `href`?

**Comment by frankieroberto** [comment]
> * The Nunnucks options for the [Phase banner component](https://design-system.service.gov.uk/components/phase-banner/) shows this for its corresponding `tag` option:
>   > Options for the tag component. See [tag](https://design-system.service.gov.uk/components/tag/#options-tag-example).

Ooh, I’d not spotted this. @owenatgov shall we use that pattern in the task list params documentation too? I wonder whether it’s better to repeat the documentation for tag rather than to just link to it, but maybe it should be consistent?
   
>   Do we want to do the same here?
> * There are options shown for `url.text`, `url.html` and `url.classes`, but going by the examples, I don’t think the `url` option exists on items. Should this not be `href`?

This was fixed within `govuk-frontend` before it was merged, but the preview on this PR isn’t showing it as that’s based on an older pre-release.

We’ll do another updated pre-release soon I think, but it’s currently blocked by something.

**Comment by colinrotherham** [comment]
Hi @frankieroberto, would you mind rebasing this PR against [`support/5.x`](https://github.com/alphagov/govuk-design-system/tree/support/5.x)?

In preparation for pre-releasing `govuk-frontend@5` we'd like to start previewing together all the GOV.UK Design System website changes that otherwise can't go into [`main`](https://github.com/alphagov/govuk-design-system/tree/main) yet whilst it's on `govuk-frontend@4`

Thanks

**Comment by 36degrees** [comment]
> Hi @frankieroberto, would you mind rebasing this PR against [`support/5.x`](https://github.com/alphagov/govuk-design-system/tree/support/5.x)?

Minor update – that should now be `release/5.0` instead. Thanks!

**Comment by frankieroberto** [comment]
@colinrotherham @36degrees sorry for the delay (I was on holiday). This PR has now been rebased from the `release/5.0` branch, and I’ve fixed up a few minor things I spotted.

I think it’s now ready for a final review. There’s one TODO left in the content for a reference to when the component is released - I’m not sure what to do about that?

I’m happy to rebase again and squash some of the minor comments when you’re ready for this to be merged. @colinrotherham also has push access to the forked repo if that’s quicker too.

File: src/components/task-list/index.md.njk
Comment (calvin-lau-sig7) [inline src/components/task-list/index.md.njk:123]:
```suggestion
This component was created by a cross-government group using research from the existing task list pattern.
```

**Comment by calvin-lau-sig7** [comment]
Hello! I've committed a change to remove a reference to the date of release — I feel like the date itself isn't absolutely essential and we can add it back in when we know (when v5 is released).

**Comment by frankieroberto** [comment]
@colinrotherham can you think of any way to resolve the duplicate ID issue, caused by embedding the same example HTML twice in the page?

**Comment by frankieroberto** [comment]
@owenatgov this is all green now, and so should be mergeable into the `release/5.0` branch if you could do a final review?

One issue that came up when converting the existing pattern page over to using the new component is that it is essential to set `idPrefix` on the task list when there are 2 or more task list components on the page (which is common for the pattern), otherwise you get duplicate IDs which would make for invalid HTML and an accessibility issue.

We could try and mitigate this by also setting `idPrefix` on all the examples on the Task list component page, even though those aren't strictly required as each example only contains a single task list component? Alternatively we could mention this directly on the task list component page guidance?

File: src/components/task-list/index.md.njk
Comment (frankieroberto) [reply src/components/task-list/index.md.njk:113]: @CharlotteDowns thanks for spotting this - must have slipped in when we drafting! 🤦  Do you think it's better with the "if" clause at the beginning or end of the sentence?

File: src/patterns/task-list-pages/index.md
Comment (stevenjmesser) [inline src/patterns/task-list-pages/index.md:19]: Examples for other page patterns include a header and footer in the example – see the [service unavailable pattern](https://design-system.service.gov.uk/patterns/service-unavailable-pages/). It’s worth adding those to remain consistent.

File: src/patterns/task-list-pages/index.md
Comment (stevenjmesser) [inline src/patterns/task-list-pages/index.md:65]: This example isn’t consistent with the guidance in the task list component guidance (or the pattern guidance) because it shows `Not started`, `In progress` and `Cannot start yet` in black, which is reserved for the `Completed` status.

**Review by stevenjmesser — CHANGES_REQUESTED** [review]
I’ve reviewed this but spotted some things that look like inconsistencies to me. Let me know if I’ve got anything wrong or don’t have the full context though!

File: src/patterns/task-list-pages/index.md
Comment (frankieroberto) [reply src/patterns/task-list-pages/index.md:19]: Good point. Will fix.

Comment (frankieroberto) [reply src/patterns/task-list-pages/index.md:19]: @stevenjmesser Fixed in 82373e3b2ca54e43ca571e57bde18403e22d5781 thanks

File: src/patterns/task-list-pages/index.md
Comment (frankieroberto) [reply src/patterns/task-list-pages/index.md:62]: Fixed in 9aac0983dd792a0395fe9cd4b86330666ee1f964 (forgot to update this example when we changed how the css classes worked, opps).

File: src/patterns/task-list-pages/index.md
Comment (frankieroberto) [inline src/patterns/task-list-pages/index.md:54]: We’ve added a lot more guidance on task list status within the component guidance page - including the suggestion that services start with just 2 statuses ("Completed" and "Incomplete") but could use 3 ("Completed", "Not yet started" and "In progress") if they find a need for it, or use custom statuses if they have additional needs.

We’ll need to decide whether to duplicate some of that content here, or only have this on either the component or pattern pages, with a link from the other one?

Any thoughts @stevenjmesser @Ciandelle @calvin-lau-sig7?

File: src/components/task-list/all-colours/index.njk
Comment (colinrotherham) [inline src/components/task-list/all-colours/index.njk:15]: Another super minor one (and I wish we had Prettier for Nunjucks) but whilst you're waiting for content replies do you mind double checking the Nunjucks indentation please?

Thanks @frankieroberto

```suggestion
      title: {
        text: "Task A"
      },
```

File: src/components/task-list/all-colours/index.njk
Comment (colinrotherham) [inline src/components/task-list/all-colours/index.njk:131]:
```suggestion
}) }}
```

File: src/components/task-list/default/index.njk
Comment (colinrotherham) [inline src/components/task-list/default/index.njk:80]:
```suggestion
  ]
}) }}
```

File: src/components/task-list/index.md.njk
Comment (colinrotherham) [inline src/components/task-list/index.md.njk:37]:
```suggestion
{{ example({ group: "components", item: "task-list", example: "default", html: true, nunjucks: true, open: false, titleSuffix: "second" }) }}
```

File: src/components/task-list/index.md.njk
Comment (colinrotherham) [inline src/components/task-list/index.md.njk:14]: Nice extra for this one

We use iframe `loading="lazy"` by default but set the first example to `loading="eager"`

```suggestion
{{ example({ group: "components", item: "task-list", example: "default", id: "default-1", html: true, nunjucks: true, open: false, loading: "eager" }) }}
```

File: src/components/task-list/index.md.njk
Comment (colinrotherham) [reply src/components/task-list/index.md.njk:37]: I'll leave you to fix the others, cheers @frankieroberto

File: src/patterns/task-list-pages/default/index.njk
Comment (colinrotherham) [inline src/patterns/task-list-pages/default/index.njk:117]: Mind using the much-forgotten `govuk-button-group` but check your margins are alright?

```suggestion
      <div class="govuk-button-group">
        {{ govukButton({
          text: "Apply"
        }) }}
      </div>
```

Comment (frankieroberto) [reply src/patterns/task-list-pages/default/index.njk]: Good shout! I forgot about that too...

File: src/components/task-list/cannot-start-yet/index.njk
Comment (colinrotherham) [inline src/components/task-list/cannot-start-yet/index.njk:15]:
```suggestion
      title: {
        text: "Company details"
      },
```

File: src/components/task-list/cannot-start-yet/index.njk
Comment (colinrotherham) [inline src/components/task-list/cannot-start-yet/index.njk:25]:
```suggestion
      title: {
        text: "Project details"
      },
```

File: src/components/task-list/cannot-start-yet/index.njk
Comment (colinrotherham) [inline src/components/task-list/cannot-start-yet/index.njk:38]:
```suggestion
      title: {
        text: "Fund applied to"
      },
```

File: src/components/task-list/cannot-start-yet/index.njk
Comment (colinrotherham) [inline src/components/task-list/cannot-start-yet/index.njk:51]:
```suggestion
  ]
}) }}
```

File: src/components/task-list/default/index.njk
Comment (colinrotherham) [inline src/components/task-list/default/index.njk:25]:
```suggestion
      title: {
        text: "Registered company details"
      },
```

File: src/components/task-list/error/index.njk
Comment (colinrotherham) [inline src/components/task-list/error/index.njk:61]:
```suggestion
  ]
}) }}
```

File: src/components/task-list/in-progress/index.njk
Comment (colinrotherham) [inline src/components/task-list/in-progress/index.njk:64]:
```suggestion
  ]
}) }}
```

File: src/components/task-list/index.md.njk
Comment (colinrotherham) [inline src/components/task-list/index.md.njk:6]:
```suggestion
backlogIssueId: 72
```

File: src/components/task-list/index.md.njk
Comment (colinrotherham) [inline src/components/task-list/index.md.njk:97]:
```suggestion
{{ example({ group: "components", item: "task-list", example: "cannot-start-yet", html: true, nunjucks: true, open: false }) }}
```

File: src/patterns/task-list-pages/task-list-statuses/index.njk
Comment (colinrotherham) [inline src/patterns/task-list-pages/task-list-statuses/index.njk:11]:
```suggestion
```

File: src/components/task-list/index.md.njk
Comment (colinrotherham) [inline src/components/task-list/index.md.njk:6]: @stevenjmesser Shall we add a separate GitHub issue for the component?

This one is tagged "pattern" (issue ID 72)

* https://github.com/alphagov/govuk-design-system-backlog/issues/72

But there aren't many tagged both "pattern" and "component"

https://github.com/alphagov/govuk-design-system-backlog/issues?q=is%3Aopen+is%3Aissue+label%3Acomponent+label%3Apattern

**Review by colinrotherham — APPROVED** [review]
Looks good to me (code, formatting)

Just got those content questions outstanding now + 2nd approval

>Just to let you know, we're operating at capacity currently and won't be able to review this until after Design System Day on 10–11 October. Lots of moving parts getting that organised.

Thanks again for all this and bearing with us @frankieroberto

File: src/components/task-list/index.md.njk
Comment (stevenjmesser) [reply src/components/task-list/index.md.njk:6]: Tag it as both, I think.

Comment (colinrotherham) [reply src/components/task-list/index.md.njk:6]: @frankieroberto @stevenjmesser Done ✅

File: src/patterns/task-list-pages/default/index.njk
Comment (frankieroberto) [reply src/patterns/task-list-pages/default/index.njk]: @stevenjmesser Have renamed the pattern to "Help users to... Complete multiple tasks". Let me know what you think!

**Review by stevenjmesser — APPROVED** [review]
This looks good, it'll give users a better idea of the scenario or use case in which to utilise this pattern.

File: src/patterns/task-list-pages/index.md
Comment (frankieroberto) [reply src/patterns/task-list-pages/index.md:54]: @stevenjmesser ok, I've moved the status guidance to the pattern page, and just left a short stub about statuses on the component page (with another link to the pattern page). It's not perfect yet, but have a look and let me know what you think...

File: src/components/task-list/index.md.njk
Comment (paulrobertlloyd) [inline src/components/task-list/index.md.njk:93]: Typo:

```suggestion
Statuses are now written in sentence case to make them easier to read. The ‘Completed’ task now uses black text with no background colour, which will draw more attention to tasks that require action.
```

**Review by paulrobertlloyd — COMMENTED** [review]
Feedback on the Task list component guidance page

File: src/patterns/complete-multiple-tasks/index.md
Comment (paulrobertlloyd) [inline src/patterns/complete-multiple-tasks/index.md:68]: This paragraph:

* doesn’t accurately describe the status colours shown
* doesn’t align with tag colour names
* is confusing as it mentions both text and background colours; elsewhere on this page only the background colour is referred to
* uses a different order of statuses to the preceding paragraph

```suggestion
‘Not yet started’ uses a blue background colour. ‘In progress’ uses a light blue background colour.
```

File: src/components/task-list/index.md.njk
Comment (stevenjmesser) [reply src/components/task-list/index.md.njk:67]: +1. (I have spotted other unnecessary quotes in other guidance, made a note to review and remove those!)

Comment (frankieroberto) [reply src/components/task-list/index.md.njk:67]: Good catch. This was probably my fault, sorry.

File: src/patterns/complete-multiple-tasks/index.md
Comment (paulrobertlloyd) [reply src/patterns/complete-multiple-tasks/index.md]: Perfect, thanks!

**Comment by colinrotherham** [comment]
Just made a few tidy up tweaks including:

1. Archive page for Task list pattern
2. Update all Task list pattern links to new page
3. Updated the road map to remove Task list
4. Squash various commits

Think we're ready to go 🙌

**Review by colinrotherham — APPROVED** [review]

**Pull request merged**
